cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1161 @@
+    return reinterpret_cast<T&>(storage);
+    _CCCL_DIAG_POP
+  }
+
+  _CCCL_HOST_DEVICE const T& operator*() const
+  {
+    _CCCL_DIAG_PUSH
+    _CCCL_DIAG_SUPPRESS_GCC("-Wstrict-aliasing")
+    return reinterpret_cast<const T&>(storage);
+    _CCCL_DIAG_POP
+  }
+};
+
+_CCCL_HOST_DEVICE inline PoorExpected<int> get_max_shared_memory()
+{
+  // gevtushenko promised me that I can assume that stream belongs to the currently active device
+  int device = 0;
+  auto error = CubDebug(cudaGetDevice(&device));
+  if (error != cudaSuccess)
+  {
+    return error;
+  }
+
+  int max_smem = 0;
+  error        = CubDebug(cudaDeviceGetAttribute(&max_smem, cudaDevAttrMaxSharedMemoryPerBlock, device));
+  if (error != cudaSuccess)
+  {
+    return error;
+  }
+
+  return max_smem;
+}
+
+template <bool RequiresStableAddress,
+          typename Offset,
+          typename RandomAccessIteratorTupleIn,
+          typename RandomAccessIteratorOut,
+          typename TransformOp,
+          typename PolicyHub = policy_hub<RequiresStableAddress, RandomAccessIteratorTupleIn>>
+struct dispatch_t;
+
+template <bool RequiresStableAddress,
+          typename Offset,
+          typename... RandomAccessIteratorsIn,
+          typename RandomAccessIteratorOut,
+          typename TransformOp,
+          typename PolicyHub>
+struct dispatch_t<RequiresStableAddress,
+                  Offset,
+                  ::cuda::std::tuple<RandomAccessIteratorsIn...>,
+                  RandomAccessIteratorOut,
+                  TransformOp,
+                  PolicyHub>
+{
+  static_assert(::cuda::std::is_same<Offset, ::cuda::std::int32_t>::value
+                  || ::cuda::std::is_same<Offset, ::cuda::std::int64_t>::value,
+                "cub::DeviceTransform is only tested and tuned for 32-bit or 64-bit signed offset types");
+
+  ::cuda::std::tuple<RandomAccessIteratorsIn...> in;
+  RandomAccessIteratorOut out;
+  Offset num_items;
+  TransformOp op;
+  cudaStream_t stream;
+
+  using kernel_ptr_t =
+    decltype(&transform_kernel<typename PolicyHub::max_policy,
+                               Offset,
+                               TransformOp,
+                               RandomAccessIteratorOut,
+                               THRUST_NS_QUALIFIER::try_unwrap_contiguous_iterator_t<RandomAccessIteratorsIn>...>);
+  kernel_ptr_t kernel;
+
+  static constexpr int loaded_bytes_per_iter = loaded_bytes_per_iteration<RandomAccessIteratorsIn...>();
+
+  // TODO(bgruber): I want to write tests for this but those are highly depending on the architecture we are running on?
+  template <typename ActivePolicy, std::size_t... Is>
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE auto configure_memcpy_async_kernel(cuda::std::index_sequence<Is...>)
+    -> PoorExpected<::cuda::std::tuple<THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron, kernel_ptr_t, int>>
+  {
+    // Benchmarking shows that even for a few iteration, this loop takes around 4-7 us, so should not be a concern.
```

Comment:
Question: would it be possible to use a static variable to cache the result? 

The primary reason the configuration has to be computed at runtime is because the occupancy of the kernel is only known after compilation. It should be constant during runtime, however, given a specific device + kernel pointer.

I had hoped that the occupancy calculation would be faster. A repeated execution of transform kernel on small data sizes (something that isn't fast or smart to begin with) would be bottlenecked on the occupancy calculation. I think in general, this is not a case we should optimize for (since it primarily occurs in unoptimized code and can be worked around by fusing lambdas), but if it is an easy fix, we had better get it in right now.

Reply:
Implemented caching. Still need to test it tomorrow.